cpg-concepts/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/concepts/Extensions.kt

```diff
@@ -0,0 +1,180 @@
+ */
+fun tag(body: TaggingContext.() -> Unit): TaggingContext {
+    val ctx = TaggingContext()
+    body(ctx)
+    return ctx
+}
+
+/**
+ * each: Applies the specified overlay nodes to "each" of the nodes selected. A node is first of all
+ * selected by its type [T]. Furthermore, the node selection can be further restricted by either a
+ * [namePredicate] or a general [predicate] over all the node's properties.
+ *
+ * This function only has an effect together with [with].
+ */
+inline fun <reified T : Node> TaggingContext.each(
+    namePredicate: CharSequence? = null,
+    noinline predicate: ((T) -> Boolean)? = null,
+): Selector<T> {
+    return Selector(T::class, namePredicate = namePredicate, predicate = predicate)
+}
+
+/**
+ * with: Specifies a [builder] that creates the actual overlay node. It is used to assign a single
+ * overlay node to a single selected "underlying" node.
+ */
+context(TaggingContext)
+fun <T : Node> Selector<T>.with(builder: BuilderContext<T>.() -> OverlayNode): EachContext<T> {
+    val ctx = EachContext(selector = this, builder = { listOf(builder(it)) })
+    listOfEach += ctx
+    return ctx
+}
+
+/**
+ * with: Specifies a [builder] that creates the actual overlay nodes. It is used to assign multiple
+ * overlay nodes to a single selected "underlying" node.
+ */
+context(TaggingContext)
+fun <T : Node> Selector<T>.withMultiple(
+    builder: BuilderContext<T>.() -> List<OverlayNode>
+): EachContext<T> {
+    val ctx = EachContext(selector = this, builder = builder)
+    listOfEach += ctx
+    return ctx
+}
+
+/**
+ * This class holds the context used in the [tag] DSL function. It basically contains a list of
+ * [EachContext], which represent a call to [each] within the tagging context.
+ */
+data class TaggingContext(val listOfEach: MutableList<EachContext<*>> = mutableListOf()) {
+    fun collect(
+        lattice: NodeToOverlayState,
+        state: NodeToOverlayStateElement,
+        node: Node,
+    ): List<OverlayNode> {
+        val list = mutableListOf<OverlayNode>()
+
+        for (each in listOfEach) {
+            // Check, if the node is assignable to the "each"
+            val overlay = each.collect(lattice, state, node)
+            list += overlay
+        }
+
+        return list
+    }
+}
+
+/**
+ * This class represents the context of an [each] DSL function. It contains the [selector] that is
+ * used to select a node that is brought into the [EachContext] as well as a [builder] that
+ * specifies how and which [OverlayNode] is constructed based on the selected node(s).
+ */
+@Suppress("UNCHECKED_CAST")
+data class EachContext<T : Node>(
+    val selector: Selector<T>,
+    val builder: (BuilderContext<T>) -> List<OverlayNode>,
+) {
+    fun collect(
+        lattice: NodeToOverlayState,
+        state: NodeToOverlayStateElement,
+        node: Node,
+    ): List<OverlayNode> {
+        // Check, if our selector matches
+        return if (selector(node as T)) {
```

Comment:
Can we get rid of this cast and move `T` to the signature without too much effort? Or should we actually make the cast with a check (`as?` and return the empty list if `null`?